### PR TITLE
Raytracing API rework, fence changes, utilities, advanced mapping

### PR DESCRIPTION
**Pull request by jkunstwald**
### General
- Reworked Raytracing API **(breaking)**
    - Added `arg::raytracing_pipeline_state_desc` to replace current arguments to PSO creation
    - Removed all occurences of `wchar_t` strings from the API, and most strings overall
    - Changed argument associations to only target either identifiable shaders or hitgroups
    - Removed data argument from TLAS creation
    - Removed `uploadTopLevelInstances` - added explicit buffer argument to `cmd::update_top_level`
    - Reworked `cmd::dispatch_rays` to support arbitrary shader table buffer locations
- Changed GPU fence API to respect weak vulkan submit-order guarantees **(breaking)**
    - Removed standalone `signalFenceGPU` and `waitFenceGPU`
    - Add args to `submit` for 1) fences to wait on before, and 2) fences to signal afterwards
- Added invalidation and flush range args to `mapBuffer` and `unmapBuffer` respectively
- Added `arg::graphics_pipeline_state_desc` and `arg::compute_pipeline_state_desc`, and PSO creation overloads
- Added additional swapchain VSync modes - synced on 2nd blank and unsynced (without allowing tearing)
- Added `format::none`, simplifying some API and commands **(breaking)**
- New functions in `util.hh`:
    - `get_texture_size_bytes`: compute the amount of bytes a texture will require on GPU
    - `get_texture_pixel_byte_offset`: compute the offset in bytes of a pixel in GPU texture data
    - `unswizzle_bgra_texture_data`: convert GPU-downloaded texture data in bgra format to rgba
    - `get_hlsl_constant_buffer_offset`: helper to compute offsets in constant buffer shader structs
    - helpers to interpret timestamp values
- Added allocator support for all static internal memory
- Added asserts about inconsistent shader views to both Backends
- Moved shared internals to `src/common`, renamed some helpers, removed namespaces
- Significantly optimized `util::dump_hex`

### D3D12
- Fixed shader table writes and RT PSO creation
- Verify all shader table writes based on signatures
- Verify GPU buffer accesses for CBV usage in draw/dispatch and shader table writes
- Vastly improved fatal error messages (OS error descriptions, full DRED breadcrumb listing with enum-to-string)
- Fixed several issues at shutdown, internal simplifications
- Added warnings about issue #50 including susceptibility check, added optional workaround
- Added `cmd::barrier_uav` for UAV barriers (no Vulkan equivalent currently)
- Enable PIX integration by default

### Vulkan
- Display SDK version at CMake config time and at backend launch

### CODE REVIEW

File: src/phantasm-hardware-interface/arguments.hh [hunk 1]
```diff
@@ -101,44 +104,110 @@ struct blas_element
...
+
+struct raytracing_pipeline_state_desc
+{
+    cc::span<raytracing_shader_library const> libraries;
+    cc::span<raytracing_argument_association const> argument_associations;
+    cc::span<raytracing_hit_group const> hit_groups;
+
+    unsigned max_recursion = 0;
+    unsigned max_payload_size_bytes = 0;
+    unsigned max_attribute_size_bytes;
```
Comment (Philip-Trettner) [inline src/phantasm-hardware-interface/arguments.hh]: not initialized. is this on purpose?

File: src/phantasm-hardware-interface/common/byte_util.hh [hunk 2]
```diff
@@ -2,12 +2,12 @@
...
 #include <cstddef>
 
-namespace phi::mem
+namespace phi::util
 {
 // From D3D12 sample MiniEngine
 // https://github.com/Microsoft/DirectX-Graphics-Samples/blob/master/MiniEngine/Core/Math/Common.h
 template <class T>
-[[nodiscard]] T align_up_masked(T value, size_t mask)
+constexpr T align_up_masked(T value, size_t mask)
```
Comment (Philip-Trettner) [inline src/phantasm-hardware-interface/common/byte_util.hh]: why is the nodiscard removed?

File: src/phantasm-hardware-interface/common/byte_util.hh [hunk 3]
```diff
@@ -19,7 +19,7 @@ template <class T>
 }
 
 template <class T>
-[[nodiscard]] T align_up(T value, size_t alignment)
+constexpr T align_up(T value, size_t alignment)
```
Comment (Philip-Trettner) [inline src/phantasm-hardware-interface/common/byte_util.hh]: why is the nodiscard removed?

File: src/phantasm-hardware-interface/common/sse_vec.hh [hunk 4]
```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <clean-core/macros.hh>
+
+#include <emmintrin.h>
+
+namespace phi
```
Comment (Philip-Trettner) [inline src/phantasm-hardware-interface/common/sse_vec.hh]: why is this public API of phi? is it used in some other lib as well?

File: src/phantasm-hardware-interface/types.hh [hunk 5]
```diff
@@ -164,16 +171,21 @@ enum class format : uint8_t
...
-[[nodiscard]] constexpr bool is_view_format(format fmt) { return fmt >= format::r24un_g8t && fmt < format::depth32f; }
+constexpr bool is_view_format(format fmt) { return fmt >= format::r24un_g8t && fmt < format::depth32f; }
 
 /// returns true if the format is a depth OR depth stencil format
-[[nodiscard]] constexpr bool is_depth_format(format fmt) { return fmt >= format::depth32f; }
+constexpr bool is_depth_format(format fmt) { return fmt >= format::depth32f; }
 
 /// returns true if the format is a depth stencil format
-[[nodiscard]] constexpr bool is_depth_stencil_format(format fmt) { return fmt >= format::depth32f_stencil8u; }
+constexpr bool is_depth_stencil_format(format fmt) { return fmt >= format::depth32f_stencil8u; }
```
Comment (Philip-Trettner) [inline src/phantasm-hardware-interface/types.hh:188]: why is nodiscard removed?

File: src/phantasm-hardware-interface/arguments.hh (hunk 1 above)
Comment (jkunstwald) [reply src/phantasm-hardware-interface/arguments.hh]: fixed

File: src/phantasm-hardware-interface/common/sse_vec.hh (hunk 4 above)
Comment (jkunstwald) [reply src/phantasm-hardware-interface/common/sse_vec.hh]: moved to phi::util

**Review by Philip-Trettner — APPROVED** [review]

**Pull request merged**